app/src/main/cpp/skyline/services/nvdrv/INvDrvServices.cpp

```diff
@@ -0,0 +1,121 @@
+        auto buffer = request.inputBuf.at(0);
+        auto path = state.process->GetString(buffer.address, buffer.size);
+        response.Push<u32>(OpenDevice(path));
+        response.Push<u32>(constant::status::Success);
+    }
+
+    void INvDrvServices::Ioctl(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto fd = request.Pop<u32>();
+        auto cmd = request.Pop<u32>();
+        state.logger->Debug("IOCTL on device: 0x{:X}, cmd: 0x{:X}", fd, cmd);
+        try {
+            if (request.inputBuf.empty() || request.outputBuf.empty()) {
+                if (request.inputBuf.empty()) {
+                    device::IoctlData data(request.outputBuf.at(0));
+                    fdMap.at(fd)->HandleIoctl(cmd, data);
+                    response.Push<u32>(data.status);
+                } else {
+                    device::IoctlData data(request.inputBuf.at(0));
+                    fdMap.at(fd)->HandleIoctl(cmd, data);
+                    response.Push<u32>(data.status);
+                }
+            } else {
+                device::IoctlData data(request.inputBuf.at(0), request.outputBuf.at(0));
+                fdMap.at(fd)->HandleIoctl(cmd, data);
+                response.Push<u32>(data.status);
+            }
+        } catch (const std::out_of_range &) {
+            throw exception("IOCTL was requested on an invalid file descriptor");
+        }
+    }
+
+    void INvDrvServices::Close(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        u32 fd = *reinterpret_cast<u32 *>(request.cmdArg);
+        state.logger->Debug("Closing NVDRV device ({})", fd);
+        try {
+            auto device = fdMap.at(fd);
+            if (!--device->refCount)
+                deviceMap.erase(device->deviceType);
+            fdMap.erase(fd);
+        } catch (const std::out_of_range &) {
+            state.logger->Warn("Trying to close non-existent FD");
+        }
+        response.Push<u32>(constant::status::Success);
+    }
+
+    void INvDrvServices::Initialize(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        response.Push<u32>(constant::status::Success);
+    }
+
+    void INvDrvServices::QueryEvent(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto fd = request.Pop<u32>();
+        auto eventId = request.Pop<u32>();
+        auto event = std::make_shared<type::KEvent>(state);
+        auto handle = state.process->InsertItem<type::KEvent>(event);
+        state.logger->Debug("QueryEvent: FD: {}, Event ID: {}, Handle: {}", fd, eventId, handle);
+        response.copyHandles.push_back(handle);
+    }
+
+    void INvDrvServices::SetAruidByPID(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        response.Push<u32>(constant::status::Success);
```

Comment:
Does u32 need to be specified? The underlying status type is u32

Reply:
I think we stick to specifying the `u32` for the results, do you think we shouldn't ?

Reply:
Well we don't do for any other enum afaik... is there any reason for needing u32? otherwise it's just a random inconsistency.